services/storehost/extStats.go

```diff
@@ -0,0 +1,434 @@
+		// }).Info("extStatsReporter: ignoring extent closed event") // #perfdisable
+
+		return true // ignore, if not any of the interesting 'intents'
+	}
+
+	// get lock shared, when reading through the map
+	t.RLock()
+	defer t.RUnlock()
+
+	ctx, ok := t.extents[string(id)]
+
+	if ok && ctx.ext == ext {
+
+		// remove ref, if it drops to zero, then send out one last report
+		if atomic.AddInt64(&ctx.ref, -1) == 0 {
+			t.trySendReport(id, ext, ctx)
+		}
+
+		// t.logger.WithFields(bark.Fields{ // #perfdisable
+		// 	common.TagExt: id,                         // #perfdisable
+		// 	`intent`:      intent,                     // #perfdisable
+		// 	`ctx.ref`:     atomic.LoadInt64(&ctx.ref), // #perfdisable
+		// }).Info("extStatsReporter: extent closed") // #perfdisable
+
+	} else {
+
+		// assert(ok) //
+		t.logger.WithFields(bark.Fields{
+			common.TagExt: id,
+			`intent`:      intent,
+			`old-ext`:     ctx.ext,
+			`new-ext`:     ext,
+		}).Error("extStatsReporter: extent-context changed unexpectedly")
+	}
+
+	return true // go ahead with the cleanup
+}
+
+func (t *ExtStatsReporter) reporterPump() {
+
+	defer t.wg.Done()
+
+	t.logger.Info("extStatsReporter: reporterPump started")
+
+pump:
+	for {
+		select {
+		case report := <-t.reportC:
+
+			var extentID, lastReport = report.extentID, report.ctx.lastReport
+
+			var lastSeqRate float64
+			if lastReport != nil {
+				lastSeqRate = common.CalculateRate(
+					common.SequenceNumber(lastReport.lastSeqNum),
+					common.SequenceNumber(report.lastSeqNum),
+					common.UnixNanoTime(lastReport.timestamp),
+					common.UnixNanoTime(report.timestamp),
+				)
+			}
+
+			var extReplStatus = shared.ExtentReplicaStatus_OPEN
+			if report.sealed {
+				extReplStatus = shared.ExtentReplicaStatus_SEALED
+			}
+
+			extReplStats := &shared.ExtentReplicaStats{
+				StoreUUID:  common.StringPtr(t.hostID),
+				ExtentUUID: common.StringPtr(extentID.String()),
+				// BeginAddress:  common.Int64Ptr(report.firstAddress),
+				// LastAddress: common.Int64Ptr(report.lastAddress),
+				BeginSequence:     common.Int64Ptr(report.firstSeqNum),
+				LastSequence:      common.Int64Ptr(report.lastSeqNum),
+				AvailableSequence: common.Int64Ptr(report.lastSeqNum),
+				// BeginEnqueueTimeUtc: common.Int64Ptr(report.firstTimestamp),
+				// LastEnqueueTimeUtc: common.Int64Ptr(report.lastTimestamp),
+				// SizeInBytes: common.Int64Ptr(report.size),
+				Status:                shared.ExtentReplicaStatusPtr(extReplStatus),
+				AvailableSequenceRate: common.Float64Ptr(lastSeqRate),
+				LastSequenceRate:      common.Float64Ptr(lastSeqRate),
+				// AvailableAddress
+				// SizeInBytesRate
+				// BeginTime
+				// EndTime
+				// WriteTime
+			}
+
+			req := &metadata.UpdateStoreExtentReplicaStatsRequest{
+				ExtentUUID:   common.StringPtr(extentID.String()),
+				ReplicaStats: []*shared.ExtentReplicaStats{extReplStats},
+			}
+
+			err := t.mClient.UpdateStoreExtentReplicaStats(nil, req)
```

Comment:
as discussed offline, there's two options here:
1. we should not expect to do this on a deleted extent -- because we update the store-extent-replica-stats only in response to an operation on store that 'mutates' the extent; and we should not expect to see any such calls onto the store after the extent has been marked for deletion.
2. if we do plan to do the check to see if it is deleted, i think it would be best to do it within the metadata-client api itself; and it's probably best to address that in a separate change.